addons/sale/models/sale_order.py

```diff
@@ -668,24 +668,50 @@ def _fetch_duplicate_orders(self):
             for order_id, duplicate_ids in result
         }
 
+    def _get_expected_date_sol_domain(self):
+        return [
+            ("order_id", "in", self.ids),
+            ("state", "!=", "cancel"),
+            ("display_type", "=", False),
+        ]
+
     @api.depends('order_line.customer_lead', 'date_order', 'state')
     def _compute_expected_date(self):
         """ For service and consumable, we only take the min dates. This method is extended in sale_stock to
             take the picking_policy of SO into account.
         """
+        max_min_customer_lead_sos = {
+            res["order_id"][0]: (res["min_customer_lead"], res["max_customer_lead"])
+            for res in self.env["sale.order.line"].read_group(
+                self._get_expected_date_sol_domain(),
+                ["order_id", "min_customer_lead:min(customer_lead)", "max_customer_lead:max(customer_lead)"],
+                ["order_id"],
+                lazy=False,
+            )
+        }
         self.mapped("order_line")  # Prefetch indication
         for order in self:
-            if order.state == 'cancel':
+            current_so_customer_lead = max_min_customer_lead_sos.get(order._origin.id)
+            if order.state == "cancel" or not current_so_customer_lead:
                 order.expected_date = False
                 continue
-            dates_list = order.order_line.filtered(
-                lambda line: not line.display_type and not line._is_delivery()
-            ).mapped(lambda line: line and line._expected_date())
+            dates_list = [order._expected_date_so(customer_lead) for customer_lead in current_so_customer_lead]
```

Comment:
I was hesitant on this, on one hand I thought of keeping the api as consistent as possible, on the other we're always choosing either the min or max of the customer lead, so I can just let odoo choose which one to choose

Do you say that it's worth it to change a bit more the methods and handle everything slightly differently? Or do we keep it more stable even in master?

Reply:
As long as we test it well and do not change the global behavior, if it's for a better and cleaner logic, it's ok.
I'd say that the biggest loss in my opinion is that we do not provide a single entry point to have the expected date anymore, since we stop relying on `_expected_date` in the compute here.